packages/amplify-category-api/src/category-utils/schema-reader.ts

```diff
@@ -0,0 +1,91 @@
+import * as fs from 'fs-extra';
+import path from 'path';
+import {
+  DocumentNode,
+  parse,
+} from 'graphql';
+import {
+  $TSAny,
+  $TSContext,
+  ApiCategoryFacade,
+} from 'amplify-cli-core';
+import { buildGraphQLTransformV2 } from '../graphql-transformer/transformer-factory';
+import {
+  SCHEMA_DIR_NAME,
+  SCHEMA_FILENAME,
+} from '../graphql-transformer/constants';
+import { generateTransformerOptions } from '../graphql-transformer/transformer-options-v2';
+import { contextUtil } from './context-util';
+
+/**
+ * SchemaReader is a utility point to consolidate and abstract GraphQL Schema reading
+ * The readSchema method provides a flag to read the un-processed (original) schema
+ * if desired, but by default the intent of the SchemaReader is to use the preProcess
+ * utility of the V2 transformer
+ */
+export class SchemaReader {
+  private schemaPath: string;
+  private schemaDocument: DocumentNode;
+  private preProcessedSchemaDocument: DocumentNode;
+
+  getSchemaPath = async (
+    resourceDir: string,
+  ): Promise<string> => {
+    if (this.schemaPath) {
+      return this.schemaPath;
+    }
+    const schemaFilePath = path.normalize(path.join(resourceDir, SCHEMA_FILENAME));
+    const schemaDirPath = path.normalize(path.join(resourceDir, SCHEMA_DIR_NAME));
+
+    if (fs.pathExistsSync(schemaFilePath)) {
+      this.schemaPath = schemaFilePath;
+    } else if (fs.pathExistsSync(schemaDirPath)) {
+      this.schemaPath = schemaDirPath;
+    } else {
+      throw new Error(`No schema found, your graphql schema should be in either ${schemaFilePath} or ${schemaDirPath}`);
+    }
+    return this.schemaPath;
+  };
+
+  invalidateCachedSchema = (): void => {
+    this.schemaPath = null;
+    this.schemaDocument = null;
+    this.preProcessedSchemaDocument = null;
+  };
+
+  readSchema = async (
+    context: $TSContext,
+    options: $TSAny,
+    usePreProcessing = true,
```

Comment:
Just want to confirm that the `schemaReader` is not being used somewhere else as this `usePreProcessing` is true by default.

Reply:
it is not, this is all new and I'm adding it without any usage currently